{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Downloading and filtering real data  \n",
    "\n",
    "For a \"real\" data test case, we'll test the different aligners/search software on the virus sequences from IMG/VR4.  \n",
    "In IMG/VR4  [DOI: 10.1093/nar/gkac1037](https://doi.org/10.1093/nar/gkac1037), the host assignment [workflow](https://github.com/jgi-microbiome-data-science/crispr-host-prediction) identifies spacer-protospacer pairs via `blastn -w 8 --dust no ...  --max_target_seqs 1000` which is roughly equivalent to using `blastn -task short` parameter, followed by filtering the results and using the `assign_host.py` script which utilizes the spacers' lineage information.  \n",
    "Note, we do not compare any filtereing done after the pairing was done, or the host assignment itself - for that reason, we also do not benchmark dedicated host assignment tools like spacerphaser, wish or iphop (even though they may be useful or utilize spacer-protospacer pairs).  \n",
    "\n",
    "In our benchmark, we'll use a more updated and curated version of the spacer set compared to the one used in IMG/VR4 - we'll use the CRISPR sapcer set from iphop [https://bitbucket.org/srouxjgi/iphop/src/main/#markdown-header-host-databases-and-versions](https://bitbucket.org/srouxjgi/iphop/src/main/#markdown-header-host-databases-and-versions). Specifically, we'll use the June 2025 upload.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fetch IMG/VR4\n",
    "The full IMG/VR4 data can be downloaded from the [JGI portal](https://genome.jgi.doe.gov/portal/pages/dynamicOrganismDownload.jsf?organism=IMG_VR). This requires logining in and selecting the files we want:  \n",
    "- IMGVR_all_nucleotides-high_confidence.fna\n",
    "- IMGVR_all_Sequence_information-high_confidence.tsv\n",
    "<!-- \n",
    "For convenience, we'll download the ffasta file from an unstable FTP hosting of them on nersc:\n",
    "- [IMGVR4_SEQUENCES.fna.zst](https://portal.nersc.gov/genomad/__data__/IMGVR_DATA/IMGVR4_SEQUENCES.fna.zst) -->\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%bash\n",
    "# mkdir imgvr4_data\n",
    "# cd imgvr4_data\n",
    "# mkdir spacers contigs\n",
    "\n",
    "# cd contigs\n",
    "# # download the contigs (this may take a while)\n",
    "# wget https://portal.nersc.gov/genomad/__data__/IMGVR_DATA/IMGVR4_SEQUENCES.fna.zst --quiet -O IMGVR4_SEQUENCES.fna.zst\n",
    "# # decompress the file\n",
    "# zstd -d IMGVR4_SEQUENCES.fna.zst\n",
    "# # get the name, length and GC% of the contigs (this may also take a while)\n",
    "# seqkit fx2tab -nl --gc IMGVR4_SEQUENCES.fna  > IMGVR4_SEQUENCES_name_length.tab"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note, we'll use the IMG_VR_2022-09-20_7.1 version of IMG/VR (v4.1, with bugfix in UViG table and protein fasta), and specifically the high-confidence genomes only ( `IMGVR_all_nucleotides-high_confidence.fna`).  \n",
    "To get the info table, you'll need to login to the JGI portal and download the `IMGVR_all_Sequence_information-high_confidence.tsv` file. The data is free and unrestricted, but still requires registerition or confirming the data use agreement. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## fetch the spacer set\n",
    "We are only interested in one (or two) files from the iphop database:\n",
    "- `All_CRISPR_spacers_nr_clean.fna`\n",
    "- `Host_Genomes.tsv`  \n",
    "\n",
    "**Note:** see the spacer_inspection.ipynb notebook for further details and analyses of the spacer set.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%bash\n",
    "# cd imgvr4_data/spacers\n",
    "# wget https://portal.nersc.gov/cfs/m342/iphop/db/extra/All_CRISPR_spacers_nr_clean.fna\n",
    "# wget https://portal.nersc.gov/cfs/m342/iphop/db/extra/Host_Genomes.tsv\n",
    "# # create a table with the name and length of the spacers\n",
    "# seqkit fx2tab -nl All_CRISPR_spacers_nr_clean.fna  > All_CRISPR_spacers_nr_clean_name_length.tab"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next some data wrangling to get the contig info and the spacer info in the same format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "contig_stats.schema: Schema([('seqid', String), ('length', Int64), ('GC', Float64)])\n",
      "contig_stats.shape: (15881302, 3)\n",
      "contig_info.schema: Schema([('UVIG', String), ('Taxon_oid', Int64), ('Scaffold_oid', String), ('coordinates', String), ('Ecosystem classification', String), ('vOTU', String), ('Length', Int64), ('Topology', String), ('geNomad score', Float64), ('Confidence', String), ('Estimated completeness', String), ('Estimated contamination', Float64), ('MIUViG quality', String), ('Gene content (total genes;cds;tRNA;geNomad marker)', String), ('Taxonomic classification', String), ('Taxonomic classification method', String), ('Host taxonomy prediction', String), ('Host prediction method', String), ('Sequence origin (doi)', String), ('seqid', String)])\n",
      "contig_info.shape: (5576197, 20)\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "os.chdir('/clusterfs/jgi/scratch/science/metagen/neri/code/blits/spacer_bench/imgvr4_data/')\n",
    "import polars as pl\n",
    "pl.Config(tbl_rows=50)\n",
    "\n",
    "from bench import *\n",
    "from bench.utils.functions import *\n",
    "\n",
    "contigs_file = 'contigs/IMGVR4_SEQUENCES.fna'\n",
    "\n",
    "contig_stats = pl.read_csv('contigs/IMGVR4_SEQUENCES_name_length.tab', separator='\\t',has_header=False,new_columns=['seqid','length',\"GC\"])\n",
    "contig_info = pl.read_csv('contigs/IMGVR_all_Sequence_information-high_confidence.tsv', separator='\\t',infer_schema_length=10000, null_values=['N/A'])\n",
    "contig_info= contig_info.rename({\"Coordinates ('whole' if the UViG is the entire contig)\":\"coordinates\"})\n",
    "contig_info= contig_info.with_columns(\n",
    "    pl.when(pl.col('coordinates').eq('whole')).then(None).otherwise(pl.col('coordinates')).alias('coordinates')\n",
    ")\n",
    "contig_info = contig_info.with_columns(\n",
    "    pl.concat_str([pl.col('UVIG'),pl.col('Taxon_oid'),pl.col('Scaffold_oid'),pl.col('coordinates')],ignore_nulls=True,separator='|').alias('seqid')\n",
    ")\n",
    "print(f\"contig_stats.schema: {contig_stats.schema}\")\n",
    "print(f\"contig_stats.shape: {contig_stats.shape}\")\n",
    "\n",
    "print(f\"contig_info.schema: {contig_info.schema}\")\n",
    "print(f\"contig_info.shape: {contig_info.shape}\")\n",
    "contig_stats = contig_stats.join(contig_info, on='seqid', how='inner')\n",
    "\n",
    "contig_stats = contig_stats.filter(pl.col(\"length\") == pl.col(\"Length\")).drop(\"Length\")\n",
    "contig_stats = contig_stats.filter(pl.col(\"Topology\") != \"GVMAG\")\n",
    "\n",
    "contig_stats = contig_stats.rename(str.lower)\n",
    "contig_stats = contig_stats.rename(lambda x: x.lower().replace(\" \",\"_\"))\n",
    "\n",
    "contig_stats = contig_stats.filter(~pl.col(\"taxonomic_classification\").str.contains(\";;;;;\"))\n",
    "contig_stats = contig_stats.with_columns(\n",
    "    pl.when(pl.col(\"miuvig_quality\").str.contains(\"High-quality|Reference\")).then(True).otherwise(False).alias(\"hq\")\n",
    ")\n",
    "contig_stats = contig_stats.with_columns([\n",
    "    pl.col(\"taxonomic_classification\")\n",
    "    .str.extract(r\"r__([^;]*)\", 1).alias(\"realm\"),\n",
    "    pl.col(\"taxonomic_classification\")\n",
    "    .str.extract(r\"k__([^;]*)\", 1).alias(\"kingdom\"),\n",
    "    pl.col(\"taxonomic_classification\")\n",
    "    .str.extract(r\"p__([^;]*)\", 1).alias(\"phylum\"),\n",
    "    pl.col(\"taxonomic_classification\")\n",
    "    .str.extract(r\"c__([^;]*)\", 1).alias(\"class\"),\n",
    "    pl.col(\"taxonomic_classification\")\n",
    "    .str.extract(r\"o__([^;]*)\", 1).alias(\"order\"),\n",
    "    pl.col(\"taxonomic_classification\")\n",
    "    .str.extract(r\"f__([^;]*)\", 1).alias(\"family\"),\n",
    "    pl.col(\"taxonomic_classification\")\n",
    "    .str.extract(r\"g__([^;]*)\", 1).alias(\"genus\"),\n",
    "    pl.col(\"taxonomic_classification\")\n",
    "    .str.extract(r\"s__([^;]*)\", 1).alias(\"species\"),\n",
    "])\n",
    "contig_stats = contig_stats.drop([\"taxonomic_classification\",\"miuvig_quality\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Removing most eukaryotic virus families"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5457198\n",
      "5115930\n"
     ]
    }
   ],
   "source": [
    "ICTV_Euk_virus_order = ['Reovirales', 'Sobelivirales', 'Patatavirales', 'Picornavirales', 'Chitovirales', 'Tolivirales', 'Ortervirales', 'Rowavirales', 'Asfuvirales', 'Ourlivirales', 'Goujianvirales', 'Nodamuvirales', 'Tymovirales', 'Herpesvirales', 'Piccovirales', 'Amarillovirales', 'Bunyavirales', 'Rohanvirales', 'Baphyvirales', 'Stellavirales', 'Algavirales', 'Cirlivirales', 'Martellivirales', 'Recrevirales', 'Pimascovirales', 'Mononegavirales', 'Zurhausenvirales', 'Mulpavirales', 'Yadokarivirales', 'Blubervirales', 'Polivirales', 'Serpentovirales', 'Ghabrivirales', 'Lefavirales', 'Cryppavirales', 'Nidovirales', 'Articulavirales', 'Wolframvirales', 'Sepolyvirales', 'Imitervirales', 'Hepelivirales', 'Muvirales', 'Jingchuvirales']\n",
    "ICTV_Euk_virus_family = [\"Abyssoviridae\", \"Adamaviridae\", \"Adenoviridae\", \"Aliusviridae\", \"Alloherpesviridae\", \"Allomimiviridae\", \"Alphaflexiviridae\", \"Alphaormycoviridae\", \"Alphasatellitidae\", \"Alphatetraviridae\", \"Alternaviridae\", \"Alvernaviridae\", \"Amalgaviridae\", \"Amesuviridae\", \"Amnoonviridae\", \"Anelloviridae\", \"Anicreviridae\", \"Arenaviridae\", \"Arteriviridae\", \"Artiviridae\", \"Artoviridae\", \"Ascoviridae\", \"Asfarviridae\", \"Aspiviridae\", \"Astroviridae\", \"Avsunviroidae\", \"Bacilladnaviridae\", \"Baculoviridae\", \"Barnaviridae\", \"Belpaoviridae\", \"Benyviridae\", \"Betaflexiviridae\", \"Betaormycoviridae\", \"Bidnaviridae\", \"Birnaviridae\", \"Bornaviridae\", \"Botourmiaviridae\", \"Botybirnaviridae\", \"Bromoviridae\", \"Burtonviroviridae\", \"Carmotetraviridae\", \"Caulimoviridae\", \"Chrysoviridae\", \"Chuviridae\", \"Closteroviridae\", \"Circoviridae\",\"Coronaviridae\", \"Cremegaviridae\", \"Crepuscuviridae\", \"Cruliviridae\", \"Curvulaviridae\", \"Deltaflexiviridae\", \"Deltanormycoviridae\", \"Dicistroviridae\", \"Discoviridae\", \"Dishuiviroviridae\", \"Draupnirviridae\", \"Dumbiviridae\", \"Endolinaviridae\", \"Endornaviridae\", \"Eupolintoviridae\", \"Euroniviridae\", \"Filamentoviridae\", \"Filoviridae\", \"Fimoviridae\", \"Flaviviridae\", \"Fusagraviridae\", \"Fusariviridae\", \"Gammaflexiviridae\", \"Gammaormycoviridae\", \"Gandrviridae\",  \"Geplanaviridae\", \"Giardiaviridae\", \"Gresnaviridae\", \"Gulliviroviridae\", \"Hadakaviridae\", \"Hantaviridae\", \"Hepadnaviridae\", \"Hepeviridae\", \"Hydriviridae\", \"Hypoviridae\", \"Hytrosaviridae\", \"Iflaviridae\", \"Inseviridae\", \"Iridoviridae\", \"Kanorauviridae\", \"Kirkoviridae\", \"Kitaviridae\", \"Kolmioviridae\", \"Konkoviridae\", \"Lebotiviridae\", \"Leishbuviridae\", \"Lispiviridae\", \"Mahapunaviridae\", \"Malacoherpesviridae\", \"Mamonoviridae\", \"Marnaviridae\", \"Marseilleviridae\", \"Matonaviridae\", \"Maviroviridae\", \"Mayoviridae\", \"Medioniviridae\", \"Megabirnaviridae\", \"Megatotiviridae\", \"Mesomimiviridae\", \"Mesoniviridae\", \"Metaviridae\", \"Metaxyviridae\", \"Mimiviridae\", \"Mitoviridae\", \"Monocitiviridae\", \"Mononiviridae\", \"Mycoalphaviridae\", \"Mymonaviridae\", \"Mypoviridae\", \"Myriaviridae\", \"Nairoviridae\", \"Nanghoshaviridae\", \"Nanhypoviridae\", \"Nanoviridae\", \"Narnaviridae\", \"Naryaviridae\", \"Natareviridae\", \"Nenyaviridae\", \"Nimaviridae\", \"Nodaviridae\", \"Noraviridae\", \"Nudiviridae\", \"Nyamiviridae\", \"Olifoviridae\", \"Omnilimnoviroviridae\", \"Oomyviridae\", \"Ootiviridae\", \"Orpheoviridae\", \"Orthoherpesviridae\", \"Orthomyxoviridae\", \"Orthototiviridae\", \"Ouroboviridae\", \"Pamosaviridae\", \"Papillomaviridae\", \"Paramyxoviridae\", \"Parvoviridae\", \"Pecoviridae\", \"Peribunyaviridae\", \"Permutotetraviridae\", \"Phasmaviridae\", \"Phenuiviridae\", \"Phlegiviridae\", \"Phycodnaviridae\", \"Phypoliviridae\", \"Picornaviridae\", \"Pistolviridae\", \"Pithoviridae\", \"Pneumoviridae\", \"Polycipiviridae\", \"Polydnaviriformidae\", \"Polymycoviridae\", \"Polyomaviridae\", \"Pospiviroidae\", \"Potyviridae\", \"Poxviridae\", \"Pseudototiviridae\", \"Pseudoviridae\", \"Qinviridae\", \"Quadriviridae\", \"Quambiviridae\", \"Redondoviridae\", \"Retroviridae\", \"Rhabdoviridae\", \"Roniviridae\", \"Ruviroviridae\", \"Sarthroviridae\", \"Schizomimiviridae\", \"Secoviridae\", \"Sedoreoviridae\", \"Sinhaliviridae\", \"Solemoviridae\", \"Solinviviridae\", \"Spiciviridae\", \"Spinareoviridae\", \"Splipalmiviridae\", \"Sputniviroviridae\", \"Sunviridae\", \"Tobaniviridae\", \"Togaviridae\", \"Tolecusatellitidae\", \"Tombusviridae\", \"Tomosaviridae\", \"Tonesaviridae\", \"Tosoviridae\", \"Tospoviridae\", \"Trimbiviridae\", \"Tulasviridae\", \"Tymoviridae\", \"Unambiviridae\", \"Vilyaviridae\", \"Virgaviridae\", \"Wupedeviridae\", \"Xinmoviridae\", \"Yadokariviridae\", \"Yadonushiviridae\", \"Yaraviridae\", \"Yueviridae\"]\n",
    "ICTV_Euk_virus_classes = ['Alsuviricetes', 'Amabiliviricetes', 'Chrymotiviricetes', 'Chunqiuviricetes','Ellioviricetes', 'Flasuviricetes', 'Herviviricetes', 'Howeltoviricetes', 'Insthoviricetes', 'Laserviricetes', 'Magsaviricetes', 'Maveriviricetes', 'Megaviricetes', 'Miaviricetes', 'Milneviricetes', 'Monjiviricetes', 'Mouviricetes', 'Naldaviricetes', 'Papovaviricetes', 'Pisoniviricetes', 'Pokkesviricetes', 'Polintoviricetes', 'Quintoviricetes', 'Repensiviricetes', 'Resentoviricetes', 'Stelpaviricetes', 'Tolucaviricetes', 'Yunchangviricetes']\n",
    "ICTV_Euk_virus_classes = set(ICTV_Euk_virus_classes).difference([\"Tectiliviricetes\",\"Arfiviricetes\",\"Revtraviricetes\"])\n",
    "\n",
    "print(contig_stats.height)\n",
    "contig_stats = contig_stats.filter(~pl.col(\"family\").is_in(set(ICTV_Euk_virus_family),nulls_equal=True))\n",
    "contig_stats = contig_stats.filter(~pl.col(\"order\").is_in(set(ICTV_Euk_virus_order),nulls_equal=True))\n",
    "contig_stats = contig_stats.filter(~pl.col(\"class\").is_in(set(ICTV_Euk_virus_classes),nulls_equal=True))\n",
    "\n",
    "print(contig_stats.height)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "realm: shape: (5, 2)\n",
      "┌───────────────┬─────────┐\n",
      "│ realm         ┆ count   │\n",
      "│ ---           ┆ ---     │\n",
      "│ str           ┆ u64     │\n",
      "╞═══════════════╪═════════╡\n",
      "│ Duplodnaviria ┆ 4986008 │\n",
      "│ Riboviria     ┆ 93153   │\n",
      "│ Monodnaviria  ┆ 35852   │\n",
      "│ Varidnaviria  ┆ 672     │\n",
      "│ Adnaviria     ┆ 245     │\n",
      "└───────────────┴─────────┘\n",
      "kingdom: shape: (8, 2)\n",
      "┌────────────────┬─────────┐\n",
      "│ kingdom        ┆ count   │\n",
      "│ ---            ┆ ---     │\n",
      "│ str            ┆ u64     │\n",
      "╞════════════════╪═════════╡\n",
      "│ Heunggongvirae ┆ 4986008 │\n",
      "│ Orthornavirae  ┆ 93153   │\n",
      "│ Sangervirae    ┆ 19063   │\n",
      "│ Loebvirae      ┆ 14198   │\n",
      "│ Shotokuvirae   ┆ 2564    │\n",
      "│ Bamfordvirae   ┆ 672     │\n",
      "│ Zilligvirae    ┆ 245     │\n",
      "│ Trapavirae     ┆ 27      │\n",
      "└────────────────┴─────────┘\n",
      "order: shape: (18, 2)\n",
      "┌───────────────────┬─────────┐\n",
      "│ order             ┆ count   │\n",
      "│ ---               ┆ ---     │\n",
      "│ str               ┆ u64     │\n",
      "╞═══════════════════╪═════════╡\n",
      "│ null              ┆ 4923798 │\n",
      "│ Crassvirales      ┆ 80669   │\n",
      "│ Timlovirales      ┆ 35701   │\n",
      "│ Norzivirales      ┆ 28300   │\n",
      "│ Petitvirales      ┆ 19063   │\n",
      "│ Tubulavirales     ┆ 14198   │\n",
      "│ Durnavirales      ┆ 10358   │\n",
      "│ Cremevirales      ┆ 1509    │\n",
      "│ Thumleimavirales  ┆ 794     │\n",
      "│ Mindivirales      ┆ 619     │\n",
      "│ Kalamavirales     ┆ 455     │\n",
      "│ Ligamenvirales    ┆ 218     │\n",
      "│ Vinavirales       ┆ 126     │\n",
      "│ Belfryvirales     ┆ 44      │\n",
      "│ Primavirales      ┆ 27      │\n",
      "│ Haloruvirales     ┆ 27      │\n",
      "│ Methanobavirales  ┆ 23      │\n",
      "│ Kirjokansivirales ┆ 1       │\n",
      "└───────────────────┴─────────┘\n",
      "class: shape: (10, 2)\n",
      "┌────────────────────┬─────────┐\n",
      "│ class              ┆ count   │\n",
      "│ ---                ┆ ---     │\n",
      "│ str                ┆ u64     │\n",
      "╞════════════════════╪═════════╡\n",
      "│ Caudoviricetes     ┆ 4986008 │\n",
      "│ Leviviricetes      ┆ 82176   │\n",
      "│ Malgrandaviricetes ┆ 19063   │\n",
      "│ Faserviricetes     ┆ 14198   │\n",
      "│ Duplopiviricetes   ┆ 10358   │\n",
      "│ Arfiviricetes      ┆ 2564    │\n",
      "│ Tectiliviricetes   ┆ 672     │\n",
      "│ Vidaverviricetes   ┆ 619     │\n",
      "│ Tokiviricetes      ┆ 245     │\n",
      "│ Huolimaviricetes   ┆ 27      │\n",
      "└────────────────────┴─────────┘\n",
      "family: shape: (59, 2)\n",
      "┌───────────────────┬─────────┐\n",
      "│ family            ┆ count   │\n",
      "│ ---               ┆ ---     │\n",
      "│ str               ┆ u64     │\n",
      "╞═══════════════════╪═════════╡\n",
      "│ null              ┆ 4915563 │\n",
      "│ Autographiviridae ┆ 42400   │\n",
      "│ Steitzviridae     ┆ 30373   │\n",
      "│ Kyanoviridae      ┆ 22808   │\n",
      "│ Microviridae      ┆ 18500   │\n",
      "│ Fiersviridae      ┆ 18298   │\n",
      "│ Demerecviridae    ┆ 15515   │\n",
      "│ Inoviridae        ┆ 12416   │\n",
      "│ Schitoviridae     ┆ 5689    │\n",
      "│ Partitiviridae    ┆ 5316    │\n",
      "│ Blumeviridae      ┆ 5261    │\n",
      "│ Atkinsviridae     ┆ 4800    │\n",
      "│ Zobellviridae     ┆ 4016    │\n",
      "│ Herelleviridae    ┆ 3615    │\n",
      "│ Picobirnaviridae  ┆ 2882    │\n",
      "│ Solspiviridae     ┆ 1874    │\n",
      "│ Straboviridae     ┆ 1746    │\n",
      "│ Ackermannviridae  ┆ 785     │\n",
      "│ Cystoviridae      ┆ 616     │\n",
      "│ Hafunaviridae     ┆ 527     │\n",
      "│ Drexlerviridae    ┆ 502     │\n",
      "│ Tectiviridae      ┆ 438     │\n",
      "│ Duinviridae       ┆ 320     │\n",
      "│ Peduoviridae      ┆ 282     │\n",
      "│ Chaseviridae      ┆ 208     │\n",
      "│ …                 ┆ …       │\n",
      "│ Salasmaviridae    ┆ 39      │\n",
      "│ Vilmaviridae      ┆ 39      │\n",
      "│ Plectroviridae    ┆ 32      │\n",
      "│ Zierdtviridae     ┆ 28      │\n",
      "│ Pleolipoviridae   ┆ 25      │\n",
      "│ Leisingerviridae  ┆ 22      │\n",
      "│ Lipothrixviridae  ┆ 17      │\n",
      "│ Duneviridae       ┆ 16      │\n",
      "│ Orlajensenviridae ┆ 11      │\n",
      "│ Guelinviridae     ┆ 10      │\n",
      "│ Tristromaviridae  ┆ 7       │\n",
      "│ Autolykiviridae   ┆ 6       │\n",
      "│ Pachyviridae      ┆ 5       │\n",
      "│ Winoviridae       ┆ 4       │\n",
      "│ Steigviridae      ┆ 4       │\n",
      "│ Assiduviridae     ┆ 3       │\n",
      "│ Forsetiviridae    ┆ 2       │\n",
      "│ Pervagoviridae    ┆ 2       │\n",
      "│ Molycolviridae    ┆ 2       │\n",
      "│ Intestiviridae    ┆ 1       │\n",
      "│ Haloferuviridae   ┆ 1       │\n",
      "│ Suoliviridae      ┆ 1       │\n",
      "│ Helgolandviridae  ┆ 1       │\n",
      "│ Naomviridae       ┆ 1       │\n",
      "│ Aggregaviridae    ┆ 1       │\n",
      "└───────────────────┴─────────┘\n",
      "genus: shape: (1_440, 2)\n",
      "┌────────────────────┬─────────┐\n",
      "│ genus              ┆ count   │\n",
      "│ ---                ┆ ---     │\n",
      "│ str                ┆ u64     │\n",
      "╞════════════════════╪═════════╡\n",
      "│ null               ┆ 5096422 │\n",
      "│ Espurtavirus       ┆ 3838    │\n",
      "│ Biseptimavirus     ┆ 3293    │\n",
      "│ Triavirus          ┆ 1444    │\n",
      "│ Alphapartitivirus  ┆ 1272    │\n",
      "│ Gammapartitivirus  ┆ 702     │\n",
      "│ Betapartitivirus   ┆ 206     │\n",
      "│ Deltapartitivirus  ┆ 187     │\n",
      "│ Fromanvirus        ┆ 183     │\n",
      "│ Casadabanvirus     ┆ 178     │\n",
      "│ Cheoctovirus       ┆ 146     │\n",
      "│ Oslovirus          ┆ 126     │\n",
      "│ Lambdavirus        ┆ 119     │\n",
      "│ Sholavirus         ┆ 117     │\n",
      "│ Skunavirus         ┆ 103     │\n",
      "│ Bicehmovirus       ┆ 102     │\n",
      "│ Tigrvirus          ┆ 98      │\n",
      "│ Tequatrovirus      ┆ 87      │\n",
      "│ Muvirus            ┆ 75      │\n",
      "│ Versovirus         ┆ 74      │\n",
      "│ Pahexavirus        ┆ 71      │\n",
      "│ Sincthavirus       ┆ 68      │\n",
      "│ Konmavirus         ┆ 67      │\n",
      "│ Traversvirus       ┆ 67      │\n",
      "│ Peduovirus         ┆ 65      │\n",
      "│ …                  ┆ …       │\n",
      "│ Nyceiraevirus      ┆ 1       │\n",
      "│ Imvubuvirus        ┆ 1       │\n",
      "│ Gammapleolipovirus ┆ 1       │\n",
      "│ Cinunavirus        ┆ 1       │\n",
      "│ Pollockvirus       ┆ 1       │\n",
      "│ Borockvirus        ┆ 1       │\n",
      "│ Gammatectivirus    ┆ 1       │\n",
      "│ Moturavirus        ┆ 1       │\n",
      "│ Astrithrvirus      ┆ 1       │\n",
      "│ Stopalavirus       ┆ 1       │\n",
      "│ Oinezvirus         ┆ 1       │\n",
      "│ Typhavirus         ┆ 1       │\n",
      "│ Spbetavirus        ┆ 1       │\n",
      "│ Jarrellvirus       ┆ 1       │\n",
      "│ Galenevirus        ┆ 1       │\n",
      "│ Nochtlivirus       ┆ 1       │\n",
      "│ Otagovirus         ┆ 1       │\n",
      "│ Reginaelenavirus   ┆ 1       │\n",
      "│ Bingvirus          ┆ 1       │\n",
      "│ Macdonaldcampvirus ┆ 1       │\n",
      "│ Huelvavirus        ┆ 1       │\n",
      "│ Xooduovirus        ┆ 1       │\n",
      "│ Harrekavirus       ┆ 1       │\n",
      "│ Hawkeyevirus       ┆ 1       │\n",
      "│ Gaiavirus          ┆ 1       │\n",
      "└────────────────────┴─────────┘\n",
      "species: shape: (1_002, 2)\n",
      "┌─────────────────────────────────────────────────────┬─────────┐\n",
      "│ species                                             ┆ count   │\n",
      "│ ---                                                 ┆ ---     │\n",
      "│ str                                                 ┆ u64     │\n",
      "╞═════════════════════════════════════════════════════╪═════════╡\n",
      "│ null                                                ┆ 5114854 │\n",
      "│ Brizovirus syn33                                    ┆ 6       │\n",
      "│ Lowelvirus tuscon4d                                 ┆ 5       │\n",
      "│ Fusarium poae virus 1                               ┆ 4       │\n",
      "│ Ceratocystis resinifera virus 1                     ┆ 4       │\n",
      "│ Pseudomonas virus Ab18                              ┆ 3       │\n",
      "│ Chalconvirus acg2014e                               ┆ 3       │\n",
      "│ Alisovirus socal22                                  ┆ 3       │\n",
      "│ Aspergillus ochraceous virus                        ┆ 3       │\n",
      "│ Beet cryptic virus 2                                ┆ 3       │\n",
      "│ Gremmeniella abietina RNA virus MS1                 ┆ 3       │\n",
      "│ Penicillium stoloniferum virus F                    ┆ 2       │\n",
      "│ Heterobasidion partitivirus 8                       ┆ 2       │\n",
      "│ Heterobasidion partitivirus 15                      ┆ 2       │\n",
      "│ Pepper cryptic virus 2                              ┆ 2       │\n",
      "│ Pepper cryptic virus 1                              ┆ 2       │\n",
      "│ Atkinsonella hypoxylon virus                        ┆ 2       │\n",
      "│ Primula malacoides virus 1                          ┆ 2       │\n",
      "│ Beet cryptic virus 1                                ┆ 2       │\n",
      "│ Cherry chlorotic rusty spot associated partitivirus ┆ 2       │\n",
      "│ Pleurotus ostreatus virus 1                         ┆ 2       │\n",
      "│ Porprismacovirus bovas2                             ┆ 2       │\n",
      "│ Dill cryptic virus 2                                ┆ 2       │\n",
      "│ Vicia cryptic virus                                 ┆ 2       │\n",
      "│ Carrot cryptic virus                                ┆ 2       │\n",
      "│ …                                                   ┆ …       │\n",
      "│ Foxunavirus fox3                                    ┆ 1       │\n",
      "│ Copernicusvirus AE417                               ┆ 1       │\n",
      "│ Camvirus joe                                        ┆ 1       │\n",
      "│ Galenevirus mbcm1                                   ┆ 1       │\n",
      "│ Chalconvirus acg2014i                               ┆ 1       │\n",
      "│ Pahexavirus PHL082M03                               ┆ 1       │\n",
      "│ Escherichia virus slur04                            ┆ 1       │\n",
      "│ Tequatrovirus ecomufv133                            ┆ 1       │\n",
      "│ Pbunavirus PaGU11                                   ┆ 1       │\n",
      "│ Cheoctovirus olympiasaint                           ┆ 1       │\n",
      "│ Janusvirus janus                                    ┆ 1       │\n",
      "│ Wizardvirus vandewege                               ┆ 1       │\n",
      "│ Macdonaldcampvirus SB28                             ┆ 1       │\n",
      "│ Rosenblumvirus portland                             ┆ 1       │\n",
      "│ Cheoctovirus blexus                                 ┆ 1       │\n",
      "│ Cheoctovirus eish                                   ┆ 1       │\n",
      "│ Benedictvirus bluefalcon                            ┆ 1       │\n",
      "│ Pbunavirus Epa7                                     ┆ 1       │\n",
      "│ Xooduovirus Xoosp2                                  ┆ 1       │\n",
      "│ Bordetella virus CN1                                ┆ 1       │\n",
      "│ Harrekavirus harreka                                ┆ 1       │\n",
      "│ Hawkeyevirus hawkeye                                ┆ 1       │\n",
      "│ Gaiavirus gaia                                      ┆ 1       │\n",
      "│ Camvirus amela                                      ┆ 1       │\n",
      "│ Cheoctovirus dante                                  ┆ 1       │\n",
      "└─────────────────────────────────────────────────────┴─────────┘\n",
      "hq: shape: (2, 2)\n",
      "┌───────┬─────────┐\n",
      "│ hq    ┆ count   │\n",
      "│ ---   ┆ ---     │\n",
      "│ bool  ┆ u64     │\n",
      "╞═══════╪═════════╡\n",
      "│ false ┆ 4694499 │\n",
      "│ true  ┆ 421431  │\n",
      "└───────┴─────────┘\n",
      "length: shape: (9, 2)\n",
      "┌────────────┬──────────────┐\n",
      "│ statistic  ┆ value        │\n",
      "│ ---        ┆ ---          │\n",
      "│ str        ┆ f64          │\n",
      "╞════════════╪══════════════╡\n",
      "│ count      ┆ 5.11593e6    │\n",
      "│ null_count ┆ 0.0          │\n",
      "│ mean       ┆ 15309.512735 │\n",
      "│ std        ┆ 19301.036966 │\n",
      "│ min        ┆ 619.0        │\n",
      "│ 25%        ┆ 5234.0       │\n",
      "│ 50%        ┆ 8230.0       │\n",
      "│ 75%        ┆ 17270.0      │\n",
      "│ max        ┆ 711471.0     │\n",
      "└────────────┴──────────────┘\n",
      "gc: shape: (9, 2)\n",
      "┌────────────┬───────────┐\n",
      "│ statistic  ┆ value     │\n",
      "│ ---        ┆ ---       │\n",
      "│ str        ┆ f64       │\n",
      "╞════════════╪═══════════╡\n",
      "│ count      ┆ 5.11593e6 │\n",
      "│ null_count ┆ 0.0       │\n",
      "│ mean       ┆ 44.819491 │\n",
      "│ std        ┆ 10.256002 │\n",
      "│ min        ┆ 17.74     │\n",
      "│ 25%        ┆ 36.53     │\n",
      "│ 50%        ┆ 42.73     │\n",
      "│ 75%        ┆ 51.95     │\n",
      "│ max        ┆ 78.41     │\n",
      "└────────────┴───────────┘\n",
      "new schema: Schema([('seqid', String), ('realm', String), ('kingdom', String), ('order', String), ('class', String), ('family', String), ('genus', String), ('species', String), ('length', Int64), ('gc', Float64), ('hq', Boolean)])\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (3, 11)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>seqid</th><th>realm</th><th>kingdom</th><th>order</th><th>class</th><th>family</th><th>genus</th><th>species</th><th>length</th><th>gc</th><th>hq</th></tr><tr><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>i64</td><td>f64</td><td>bool</td></tr></thead><tbody><tr><td>&quot;IMGVR_UViG_3300017956_002023|3300017956|Ga0181580_10017582&quot;</td><td>&quot;Duplodnaviria&quot;</td><td>&quot;Heunggongvirae&quot;</td><td>null</td><td>&quot;Caudoviricetes&quot;</td><td>null</td><td>null</td><td>null</td><td>5634</td><td>42.76</td><td>false</td></tr><tr><td>&quot;IMGVR_UViG_3300027719_000271|3300027719|Ga0209467_1000224|56315-89830&quot;</td><td>&quot;Duplodnaviria&quot;</td><td>&quot;Heunggongvirae&quot;</td><td>null</td><td>&quot;Caudoviricetes&quot;</td><td>null</td><td>null</td><td>null</td><td>33516</td><td>54.51</td><td>false</td></tr><tr><td>&quot;IMGVR_UViG_3300047388_000595|3300047388|Ga0497965_0013996&quot;</td><td>&quot;Duplodnaviria&quot;</td><td>&quot;Heunggongvirae&quot;</td><td>null</td><td>&quot;Caudoviricetes&quot;</td><td>null</td><td>null</td><td>null</td><td>5592</td><td>57.92</td><td>false</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (3, 11)\n",
       "┌───────────────────────────────────────────────────────────────────────┬───────────────┬────────────────┬───────┬────────────────┬────────┬───────┬─────────┬────────┬───────┬───────┐\n",
       "│ seqid                                                                 ┆ realm         ┆ kingdom        ┆ order ┆ class          ┆ family ┆ genus ┆ species ┆ length ┆ gc    ┆ hq    │\n",
       "│ ---                                                                   ┆ ---           ┆ ---            ┆ ---   ┆ ---            ┆ ---    ┆ ---   ┆ ---     ┆ ---    ┆ ---   ┆ ---   │\n",
       "│ str                                                                   ┆ str           ┆ str            ┆ str   ┆ str            ┆ str    ┆ str   ┆ str     ┆ i64    ┆ f64   ┆ bool  │\n",
       "╞═══════════════════════════════════════════════════════════════════════╪═══════════════╪════════════════╪═══════╪════════════════╪════════╪═══════╪═════════╪════════╪═══════╪═══════╡\n",
       "│ IMGVR_UViG_3300017956_002023|3300017956|Ga0181580_10017582            ┆ Duplodnaviria ┆ Heunggongvirae ┆ null  ┆ Caudoviricetes ┆ null   ┆ null  ┆ null    ┆ 5634   ┆ 42.76 ┆ false │\n",
       "│ IMGVR_UViG_3300027719_000271|3300027719|Ga0209467_1000224|56315-89830 ┆ Duplodnaviria ┆ Heunggongvirae ┆ null  ┆ Caudoviricetes ┆ null   ┆ null  ┆ null    ┆ 33516  ┆ 54.51 ┆ false │\n",
       "│ IMGVR_UViG_3300047388_000595|3300047388|Ga0497965_0013996             ┆ Duplodnaviria ┆ Heunggongvirae ┆ null  ┆ Caudoviricetes ┆ null   ┆ null  ┆ null    ┆ 5592   ┆ 57.92 ┆ false │\n",
       "└───────────────────────────────────────────────────────────────────────┴───────────────┴────────────────┴───────┴────────────────┴────────┴───────┴─────────┴────────┴───────┴───────┘"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contig_stats = contig_stats.select([\"seqid\",\"realm\",\"kingdom\",\"order\",\"class\",\"family\",\"genus\",\"species\",\"length\",\"gc\",\"hq\"])\n",
    "\n",
    "for i in [\"realm\",\"kingdom\",\"order\",\"class\",\"family\",\"genus\",\"species\",\"hq\"]:\n",
    "    print(f\"{i}: {contig_stats[i].value_counts(sort=True)}\")\n",
    "for i in [\"length\",\"gc\"]:\n",
    "    print(f\"{i}: {contig_stats[i].describe()}\")\n",
    "\n",
    "print(f\"new schema: {contig_stats.schema}\")\n",
    "contig_stats.sample(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Removing very short contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5115930\n",
      "5115894\n"
     ]
    }
   ],
   "source": [
    "print(contig_stats.height)\n",
    "contig_stats = contig_stats.filter(pl.col(\"length\") > 1000)\n",
    "print(contig_stats.height)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Export out (stats and fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using 10 threads\n",
      "Loaded 5115894 headers\n",
      "Processed 100000 records\n",
      "Processed 200000 records\n",
      "Processed 300000 records\n",
      "Processed 400000 records\n",
      "Processed 500000 records\n",
      "Processed 600000 records\n",
      "Processed 700000 records\n",
      "Processed 800000 records\n",
      "Processed 900000 records\n",
      "Processed 1000000 records\n",
      "Processed 1100000 records\n",
      "Processed 1200000 records\n",
      "Processed 1300000 records\n",
      "Processed 1400000 records\n",
      "Processed 1500000 records\n",
      "Processed 1600000 records\n",
      "Processed 1700000 records\n",
      "Processed 1800000 records\n",
      "Processed 1900000 records\n",
      "Processed 2000000 records\n",
      "Processed 2100000 records\n",
      "Processed 2200000 records\n",
      "Processed 2300000 records\n",
      "Processed 2400000 records\n",
      "Processed 2500000 records\n",
      "Processed 2600000 records\n",
      "Processed 2700000 records\n",
      "Processed 2800000 records\n",
      "Processed 2900000 records\n",
      "Processed 3000000 records\n",
      "Processed 3100000 records\n",
      "Processed 3200000 records\n",
      "Processed 3300000 records\n",
      "Processed 3400000 records\n",
      "Processed 3500000 records\n",
      "Processed 3600000 records\n",
      "Processed 3700000 records\n",
      "Processed 3800000 records\n",
      "Processed 3900000 records\n",
      "Processed 4000000 records\n",
      "Processed 4100000 records\n",
      "Processed 4200000 records\n",
      "Processed 4300000 records\n",
      "Processed 4400000 records\n",
      "Processed 4500000 records\n",
      "Processed 4600000 records\n",
      "Processed 4700000 records\n",
      "Processed 4800000 records\n",
      "Processed 4900000 records\n",
      "Processed 5000000 records\n",
      "Processed 5100000 records\n",
      "Processed 5200000 records\n",
      "Processed 5300000 records\n",
      "Processed 5400000 records\n",
      "Processed 5500000 records\n",
      "Processed 5600000 records\n",
      "Processed 5700000 records\n",
      "Processed 5800000 records\n",
      "Processed 5900000 records\n",
      "Processed 6000000 records\n",
      "Processed 6100000 records\n",
      "Processed 6200000 records\n",
      "Processed 6300000 records\n",
      "Processed 6400000 records\n",
      "Processed 6500000 records\n",
      "Processed 6600000 records\n",
      "Processed 6700000 records\n",
      "Processed 6800000 records\n",
      "Processed 6900000 records\n",
      "Processed 7000000 records\n",
      "Processed 7100000 records\n",
      "Processed 7200000 records\n",
      "Processed 7300000 records\n",
      "Processed 7400000 records\n",
      "Processed 7500000 records\n",
      "Processed 7600000 records\n",
      "Processed 7700000 records\n",
      "Processed 7800000 records\n",
      "Processed 7900000 records\n",
      "Processed 8000000 records\n",
      "Processed 8100000 records\n",
      "Processed 8200000 records\n",
      "Processed 8300000 records\n",
      "Processed 8400000 records\n",
      "Processed 8500000 records\n",
      "Processed 8600000 records\n",
      "Processed 8700000 records\n",
      "Processed 8800000 records\n",
      "Processed 8900000 records\n",
      "Processed 9000000 records\n",
      "Processed 9100000 records\n",
      "Processed 9200000 records\n",
      "Processed 9300000 records\n",
      "Processed 9400000 records\n",
      "Processed 9500000 records\n",
      "Processed 9600000 records\n",
      "Processed 9700000 records\n",
      "Processed 9800000 records\n",
      "Processed 9900000 records\n",
      "Processed 10000000 records\n",
      "Processed 10100000 records\n",
      "Processed 10200000 records\n",
      "Processed 10300000 records\n",
      "Processed 10400000 records\n",
      "Processed 10500000 records\n",
      "Processed 10600000 records\n",
      "Processed 10700000 records\n",
      "Processed 10800000 records\n",
      "Processed 10900000 records\n",
      "Processed 11000000 records\n",
      "Processed 11100000 records\n",
      "Processed 11200000 records\n",
      "Processed 11300000 records\n",
      "Processed 11400000 records\n",
      "Processed 11500000 records\n",
      "Processed 11600000 records\n",
      "Processed 11700000 records\n",
      "Processed 11800000 records\n",
      "Processed 11900000 records\n",
      "Processed 12000000 records\n",
      "Processed 12100000 records\n",
      "Processed 12200000 records\n",
      "Processed 12300000 records\n",
      "Processed 12400000 records\n",
      "Processed 12500000 records\n",
      "Processed 12600000 records\n",
      "Processed 12700000 records\n",
      "Processed 12800000 records\n",
      "Processed 12900000 records\n",
      "Processed 13000000 records\n",
      "Processed 13100000 records\n",
      "Processed 13200000 records\n",
      "Processed 13300000 records\n",
      "Processed 13400000 records\n",
      "Processed 13500000 records\n",
      "Processed 13600000 records\n",
      "Processed 13700000 records\n",
      "Processed 13800000 records\n",
      "Processed 13900000 records\n",
      "Processed 14000000 records\n",
      "Processed 14100000 records\n",
      "Processed 14200000 records\n",
      "Processed 14300000 records\n",
      "Processed 14400000 records\n",
      "Processed 14500000 records\n",
      "Processed 14600000 records\n",
      "Processed 14700000 records\n",
      "Processed 14800000 records\n",
      "Processed 14900000 records\n",
      "Processed 15000000 records\n",
      "Processed 15100000 records\n",
      "Processed 15200000 records\n",
      "Processed 15300000 records\n",
      "Processed 15400000 records\n",
      "Processed 15500000 records\n",
      "Processed 15600000 records\n",
      "Processed 15700000 records\n",
      "Processed 15800000 records\n",
      "Processed 15881302 records, written 5115894 records\n"
     ]
    }
   ],
   "source": [
    "contig_stats.write_parquet(\"./contigs/filtered_contig_stats.parquet\")\n",
    "contig_stats.select(pl.col(\"seqid\")).write_csv(\"./contigs/filtered_contig.lst\",include_header=False)\n",
    "\n",
    "!paraseq_filt --headers ./contigs/filtered_contig.lst --input contigs/IMGVR4_SEQUENCES.fna --output contigs/filtered_contigs.fasta --threads 10"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we'll index the contigs file for fast random access later.\n",
    "This may take a while, but only done once and will save time (later)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " 74G -rw-rw-r-- 1 uneri grp-org-sc-metagen  74G Jan 22 16:22 ../imgvr4_data/contigs/filtered_contigs.fasta\n",
      "767M -rw-r--r-- 1 uneri grp-org-sc-metagen 767M Jan 22 16:15 ../imgvr4_data/contigs/filtered_contigs.fasta.fxi\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "pyfastx index contigs/filtered_contigs.fasta\n",
    "ls -lsh ../imgvr4_data/contigs/filtered_contigs.fasta*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "(Next steps are in the spacer inspection notebook)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "default",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
