{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using real-data  \n",
    "\n",
    "For a \"real\" data test case, we'll test the different aligners/search software on the virus sequences from IMG/VR4.  \n",
    "In IMG/VR4  [DOI: 10.1093/nar/gkac1037](https://doi.org/10.1093/nar/gkac1037), the host assignment [workflow](https://github.com/jgi-microbiome-data-science/crispr-host-prediction) identifies spacer-protospacer pairs via `blastn -w 8 --dust no ...  --max_target_seqs 1000` which is roughly equivalent to using `blastn -task short` parameter, followed by filtering the results and using the `assign_host.py` script which utilizes the spacers' lineage information.  \n",
    "Note, we do not compare any filtereing done after the pairing was done, or the host assignment itself - for that reason, we also do not benchmark dedicated host assignment tools like spacerphaser, wish or iphop (even though they may be useful or utilize spacer-protospacer pairs).  \n",
    "\n",
    "In our benchmark, we'll use a more updated and curated version of the spacer set compared to the one used in IMG/VR4 - we'll use the CRISPR sapcer set from iphop [https://bitbucket.org/srouxjgi/iphop/src/main/#markdown-header-host-databases-and-versions](https://bitbucket.org/srouxjgi/iphop/src/main/#markdown-header-host-databases-and-versions). Specifically, we'll use the June 2025 upload.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fetch IMG/VR4\n",
    "The full IMG/VR4 data can be downloaded from the [JGI portal](https://genome.jgi.doe.gov/portal/pages/dynamicOrganismDownload.jsf?organism=IMG_VR). This requires logining in and selecting the files we want:  \n",
    "- IMGVR_all_nucleotides-high_confidence.fna\n",
    "- IMGVR_all_Sequence_information-high_confidence.tsv\n",
    "\n",
    "For convenience, we'll download the ffasta file from an unstable FTP hosting of them on nersc:\n",
    "- [IMGVR4_SEQUENCES.fna.zst](https://portal.nersc.gov/genomad/__data__/IMGVR_DATA/IMGVR4_SEQUENCES.fna.zst)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "mkdir imgvr4_data\n",
    "cd imgvr4_data\n",
    "mkdir spacers contigs\n",
    "\n",
    "cd contigs\n",
    "# download the contigs (this may take a while)\n",
    "wget https://portal.nersc.gov/genomad/__data__/IMGVR_DATA/IMGVR4_SEQUENCES.fna.zst --quiet -O IMGVR4_SEQUENCES.fna.zst\n",
    "# decompress the file\n",
    "zstd -d IMGVR4_SEQUENCES.fna.zst\n",
    "# get the name, length and GC% of the contigs (this may also take a while)\n",
    "seqkit fx2tab -nl --gc IMGVR4_SEQUENCES.fna  > IMGVR4_SEQUENCES_name_length.tab"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note, we'll use the IMG_VR_2022-09-20_7.1 version of IMG/VR (v4.1, with bugfix in UViG table and protein fasta), and specifically the high-confidence genomes only ( `IMGVR_all_nucleotides-high_confidence.fna`).  \n",
    "To get the info table, you'll need to login to the JGI portal and download the `IMGVR_all_Sequence_information-high_confidence.tsv` file. The data is free and unrestricted, but still requires registerition or confirming the data use agreement. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## fetch the spacer set\n",
    "We are only interested in one (or two) files from the iphop database:\n",
    "- `All_CRISPR_spacers_nr_clean.fna`\n",
    "- `Host_Genomes.tsv`\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cd imgvr4_data/spacers\n",
    "wget https://portal.nersc.gov/cfs/m342/iphop/db/extra/All_CRISPR_spacers_nr_clean.fna\n",
    "wget https://portal.nersc.gov/cfs/m342/iphop/db/extra/Host_Genomes.tsv\n",
    "# create a table with the name and length of the spacers\n",
    "seqkit fx2tab -nl All_CRISPR_spacers_nr_clean.fna  > All_CRISPR_spacers_nr_clean_name_length.tab"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next some data wrangling to get the contig info and the spacer info in the same format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.chdir('/clusterfs/jgi/scratch/science/metagen/neri/code/blits/spacer_bench/imgvr4_data/')\n",
    "import polars as pl\n",
    "pl.Config(tbl_rows=50)\n",
    "\n",
    "from bench import *\n",
    "from bench.utils.functions import *\n",
    "\n",
    "contigs_file = 'contigs/IMGVR4_SEQUENCES.fna'\n",
    "spacers_file = 'spacers/All_CRISPR_spacers_nr_clean.fna'\n",
    "\n",
    "contig_stats = pl.read_csv('contigs/IMGVR4_SEQUENCES_name_length.tab', separator='\\t',has_header=False,new_columns=['seqid','length',\"GC\"])\n",
    "contig_info = pl.read_csv('contigs/IMGVR_all_Sequence_information-high_confidence.tsv', separator='\\t',infer_schema_length=10000, null_values=['N/A'])\n",
    "contig_info= contig_info.rename({\"Coordinates ('whole' if the UViG is the entire contig)\":\"coordinates\"})\n",
    "contig_info= contig_info.with_columns(\n",
    "    pl.when(pl.col('coordinates').eq('whole')).then(None).otherwise(pl.col('coordinates')).alias('coordinates')\n",
    ")\n",
    "contig_info = contig_info.with_columns(\n",
    "    pl.concat_str([pl.col('UVIG'),pl.col('Taxon_oid'),pl.col('Scaffold_oid'),pl.col('coordinates')],ignore_nulls=True,separator='|').alias('seqid')\n",
    ")\n",
    "print(f\"contig_stats.schema: {contig_stats.schema}\")\n",
    "print(f\"contig_stats.shape: {contig_stats.shape}\")\n",
    "\n",
    "print(f\"contig_info.schema: {contig_info.schema}\")\n",
    "print(f\"contig_info.shape: {contig_info.shape}\")\n",
    "contig_stats = contig_stats.join(contig_info, on='seqid', how='inner')\n",
    "\n",
    "contig_stats = contig_stats.filter(pl.col(\"length\") == pl.col(\"Length\"))\n",
    "contig_stats = contig_stats.filter(pl.col(\"Topology\") != \"GVMAG\")\n",
    "\n",
    "contig_stats = contig_stats.select([\"seqid\",\"Taxonomic classification\",\"length\",\"GC\",\"Topology\",\"MIUViG quality\",\"UVIG\"])\n",
    "contig_stats = contig_stats.rename(str.lower)\n",
    "contig_stats = contig_stats.rename(lambda x: x.lower().replace(\" \",\"_\"))\n",
    "\n",
    "contig_stats = contig_stats.filter(~pl.col(\"taxonomic_classification\").str.contains(\";;;;;\"))\n",
    "contig_stats = contig_stats.with_columns(\n",
    "    pl.when(pl.col(\"miuvig_quality\").str.contains(\"High-quality|Reference\")).then(True).otherwise(False).alias(\"hq\")\n",
    ")\n",
    "contig_stats = contig_stats.with_columns([\n",
    "    pl.col(\"taxonomic_classification\")\n",
    "    .str.extract(r\"r__([^;]*)\", 1).alias(\"realm\"),\n",
    "    pl.col(\"taxonomic_classification\")\n",
    "    .str.extract(r\"k__([^;]*)\", 1).alias(\"kingdom\"),\n",
    "    pl.col(\"taxonomic_classification\")\n",
    "    .str.extract(r\"p__([^;]*)\", 1).alias(\"phylum\"),\n",
    "    pl.col(\"taxonomic_classification\")\n",
    "    .str.extract(r\"c__([^;]*)\", 1).alias(\"class\"),\n",
    "    pl.col(\"taxonomic_classification\")\n",
    "    .str.extract(r\"o__([^;]*)\", 1).alias(\"order\"),\n",
    "    pl.col(\"taxonomic_classification\")\n",
    "    .str.extract(r\"f__([^;]*)\", 1).alias(\"family\"),\n",
    "    pl.col(\"taxonomic_classification\")\n",
    "    .str.extract(r\"g__([^;]*)\", 1).alias(\"genus\"),\n",
    "    pl.col(\"taxonomic_classification\")\n",
    "    .str.extract(r\"s__([^;]*)\", 1).alias(\"species\"),\n",
    "])\n",
    "contig_stats = contig_stats.drop([\"taxonomic_classification\",\"miuvig_quality\"])\n",
    "for i in contig_stats.columns:\n",
    "    print(f\"{i}: {contig_stats[i].value_counts(sort=True)}\")\n",
    "# tmp.head(3)\n",
    "print(f\"new schema: {contig_stats.schema}\")\n",
    "contig_stats.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we'll index the contigs file for fast random access later.\n",
    "This may take a while, but only done once and will save time (later)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "158G -rw-rw-r-- 1 uneri grp-org-sc-metagen 158G Sep  4  2022 ../imgvr4_data/contigs/IMGVR4_SEQUENCES.fna\n",
      "2.3G -rw-r--r-- 1 uneri grp-org-sc-metagen 2.3G Oct  7 15:38 ../imgvr4_data/contigs/IMGVR4_SEQUENCES.fna.fxi\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "pyfastx index contigs/IMGVR4_SEQUENCES.fna\n",
    "ls -lsh ../imgvr4_data/contigs/IMGVR4_SEQUENCES.fna*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "(Next steps are in the spacer inspection notebook)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "default",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.22"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
